File: leetcode/src/main/java/leetcode/editor/cn/最小化曼哈顿距离minimizeManhattanDistances.java
/**
 * @Title: 3102.最小化曼哈顿距离
 * @TitleSlug: minimizeManhattanDistances
 * @Author: Neo
 * @Date: 2024-04-05 19:22:56
 */
package leetcode.editor.cn;

public class 最小化曼哈顿距离minimizeManhattanDistances {
    public static void main(String[] args) {
        // 测试代码
        // Solution solution = new 最小化曼哈顿距离minimizeManhattanDistances().new Solution();
    }
    
    // leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        public int minimumDistance(int[][] points) {
            final int inf = Integer.MAX_VALUE;
            int[] maxX = new int[]{-inf, -inf}, maxY = new int[]{-inf, -inf};
            int[] minX = new int[]{inf, inf}, minY = new int[]{inf, inf};
            for (int[] p : points) {
                int x = p[0] + p[1];
                int y = p[1] - p[0];
                compareMax(maxX, x);
                compareMax(maxY, y);
                compareMin(minX, x);
                compareMin(minY, y);
            }
            int ans = inf;
            for (int[] p : points) {
                int x = p[0] + p[1], y = p[1] - p[0];
                int dx = (x == maxX[0] ? maxX[1] : maxX[0]) - (x == minX[0] ? minX[1] : minX[0]);
                int dy = (y == maxY[0] ? maxY[1] : maxY[0]) - (y == minY[0] ? minY[1] : minY[0]);
                ans = Math.min(ans, Math.max(dx, dy));
            }
            return ans;
        }
        
        private void compareMax(int[] max, int i) {
            if (i > max[0]) {
                max[1] = max[0];
                max[0] = i;
            } else if (i > max[1]) {
                max[1] = i;
            }
        }
        
        private void compareMin(int[] min, int i) {
            if (i < min[0]) {
                min[1] = min[0];
                min[0] = i;
            } else if (i < min[1]) {
                min[1] = i;
            }
        }
    }
// leetcode submit region end(Prohibit modification and deletion)

}
       

File: leetcode/src/main/python/leetcode/editor/cn/最小化曼哈顿距离minimizeManhattanDistances.py
"""
Title: 3102.最小化曼哈顿距离
Title Slug: minimizeManhattanDistances
Author: Neo
Date: 2024-07-09 12:17:09
"""
from cmath import inf
from typing import List


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def minimumDistance(self, points: List[List[int]]) -> int:
        def compareMax(mx: List[int], x) -> None:
            if x > mx[0]:
                mx[0], mx[1] = x, mx[0]
            elif x > mx[1]:
                mx[1] = x

        def compareMin(mn: List[int], x) -> None:
            if x < mn[0]:
                mn[0], mn[1] = x, mn[0]
            elif x < mn[1]:
                mn[1] = x

        mxX, mxY = [-inf, -inf], [-inf, -inf]
        mnX, mnY = [inf, inf], [inf, inf]
        for (i, j) in points:
            x, y = i + j, j - i
            compareMax(mxX, x)
            compareMax(mxY, y)
            compareMin(mnX, x)
            compareMin(mnY, y)

        ans = inf
        for (i, j) in points:
            x, y = i + j, j - i
            dx = (mxX[1] if x == mxX[0] else mxX[0]) - (mnX[1] if x == mnX[0] else mnX[0])
            dy = (mxY[1] if y == mxY[0] else mxY[0]) - (mnY[1] if y == mnY[0] else mnY[0])
            ans = min(ans, max(dx, dy))
        return ans


# leetcode submit region end(Prohibit modification and deletion)
Solution().minimumDistance([[3, 10], [5, 15], [10, 2], [4, 4]])
